chipset_enable.c

```diff
@@ -1456,6 +1456,141 @@ static int enable_flash_sb600(const struct programmer_cfg *cfg, struct pci_dev *
 	return ret;
 }
 
+static int enable_flash_fch(const struct programmer_cfg *cfg, struct pci_dev *dev, const char *name)
+{
+	uint32_t prot, length;
+	uint8_t reg, spi_alt_cs;
+	int ret = 0;
+	uint8_t *fch_spi;
+	bool read_prot, write_prot;
+
+	uint32_t spibar = pci_read_long(dev, 0xa0) & 0xffffffc0; /* bits 31:6, SB600 has 31:5 */
+	if (spibar == 0) {
+		msg_perr("Error: Invalid SPI BASE\n");
+		return ERROR_FATAL;
+	}
+
+	/* Physical memory has to be mapped at page (4k) boundaries. */
+	fch_spi = rphysmap("FCH SPIBASE", spibar & 0xfffff000, 0x1000);
+	if (fch_spi == NULL) {
+		msg_perr("Error: Failed to map SPI BASE\n");
+		return ERROR_FATAL;
+	}
+
+	/* The low bits of the SPI base address are used as offset into
+	 * the mapped page.
+	 */
+	fch_spi += spibar & 0xfff;
+
+	read_prot = write_prot = false;
+
+	spi_alt_cs = mmio_readb(fch_spi + 0x1d);
+
+	/* ROM Protect registers active? */
+	if (spi_alt_cs & 0x8) {
+		/* Clear ROM protect 0-3. */
+		for (reg = 0x50; reg < 0x60; reg += 4) {
+			prot = pci_read_long(dev, reg);
+			/* No protection flags for this region?*/
+			if ((prot & 0x600) == 0)
+				continue;
+
+			length = ((prot & 0xff) + 1) << ((prot & 0x100) ? 16 : 12);
+			msg_pdbg("Chipset %s%sprotected flash from 0x%08x to 0x%08x, unlocking...",
+				  (prot & 0x200) ? "read " : "",
+				  (prot & 0x400) ? "write " : "",
+				  (prot & 0xfffff000),
+				  (prot & 0xfffff000) + (length - 1));
+
+			/* ROM Protect registers locked? */
+			if (spi_alt_cs & 0x20) {
+				read_prot = read_prot || (prot & 0x200);
+				write_prot = write_prot || (prot & 0x400);
+				msg_pdbg("FAILED!\n");
```

Comment:
Here and below are just to not report the same thing for two very different failures.
```suggestion
				msg_pdbg("FAILED! (ROM Protect registers are locked.)\n");
```